gen/gen.go

```diff
@@ -119,23 +148,67 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
 	return ioutil.WriteFile(filepath.Join(dirProblem, "cases_test.go"), src, 0666)
 }
 
-func getPath(jFile string) (jPath, jOrigin, jCommit string) {
+func getLocal(jFile string) (jPath, jOrigin, jCommit string) {
 	// Ideally draw from a .json which is pulled from the official x-common
 	// repository.  For development however, accept a file in current directory
 	// if there is no .json in source control.  Also allow an override in any
 	// case by environment variable.
-	if jPath = os.Getenv("EXTEST"); jPath > "" {
+	if jPath := os.Getenv("EXTEST"); jPath > "" {
 		return jPath, "local file", "" // override
 	}
+
 	c := exec.Command("git", "log", "-1", "--oneline", jFile)
 	c.Dir = dirMetadata
-	ori, err := c.Output()
+	origin, err := c.Output()
 	if err != nil {
 		return "", "local file", "" // no source control
 	}
 	if _, err = os.Stat(filepath.Join(c.Dir, jFile)); err != nil {
 		return "", "local file", "" // not in source control
 	}
+
 	// good.  return source control dir and commit.
-	return c.Dir, "exercism/x-common", string(bytes.TrimSpace(ori))
+	fmt.Printf("Fetching canonical data from a local source.\nHave you ensured it is up to date?\n")
+	return c.Dir, "local exercism/x-common", string(bytes.TrimSpace(origin))
+}
+
+func getRemote(exercise string) (body []byte, jOrigin string, jCommit string, err error) {
+	fmt.Println("Collecting remote data.")
+	resp, err := genClient.Get(fmt.Sprintf(canonicalDataURL, exercise))
+	if err != nil {
+		return []byte{}, "", "", err
+	}
+	if resp.Status != "200 OK" {
+		return []byte{}, "", "", fmt.Errorf("error fetching remote data: %s", resp.Status)
+	}
+	defer resp.Body.Close()
+	body, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return []byte{}, "", "", err
+	}
+	c, err := getRemoteCommit(exercise)
+	if err != nil {
+		return body, "github.com/exercism/x-common", "", nil
```

Comment:
That makes a lot of sense, it is the biggest issue with fetching from the network, especially if you have to deal with rural internet! Good reason to keep this using the local first, before resorting to the remote fetch. Thanks for clarifying this for me.